File: С++/fourth_task.cpp
#include <iostream>
#include <vector>
#include <map>

using namespace std;

char methodBord(vector<vector<char>> chains) {
    map<char,int> scores;

    for(int i = 0; i < chains[0].size(); i++) {
        scores[chains[0][i]] = 0;
    }

    for (auto& chain : chains) {
        for (int i = 0; i < chain.size(); i++) {
            scores[chain[i]] += chain.size() - 1 - i;
        }
    }

    int max = 0;
    char max_key = ' ';
    for (auto& [a, b] : scores) {
        if (b > max) {
            max_key = a;
            max = b;
        }
    }

    return max_key;
}

char methodCondorse(vector<vector<char>> chains) {
    map<char,int> scores;

    for(int i = 0; i < chains[0].size(); i++) {
        scores[chains[0][i]] = 0;
    }

    for (auto& chain : chains) {
        scores[chain[0]]++;
    }

    int max = 0;
    char max_key = ' ';
    for (auto& [a, b] : scores) {
        if (b > max) {
            max_key = a;
            max = b;
        }
    }

    return max_key;
}

int main() {
    int n, k;
    cin >> n >> k;

    vector<vector<char>> chains;
    chains.resize(k);
    for (int i = 0; i < k; i++) {
        chains[i].resize(n);
    }

    for (int i = 0; i < k; i++) {
        cout << "Цепь для " << i+1 << " избирателя:" << endl;
        for (int j = 0; j < n; j++) {
            char letter;
            cin >> letter;
            chains[i][j] = letter;
        }
        cout << endl;
    }

    cout << "По методу Борда выигрывает кандидат " << methodBord(chains) << endl;
    cout << "По методу Кондорсе выигрывает кандидат " << methodCondorse(chains) << endl;

    return 0;
}

File: Python/fourth_task.py
def method_bord(chains):
    scores = {c: 0 for c in chains[0]}

    for chain in chains:
        for i in range(len(chain)):
            scores[chain[i]] += len(chain) - 1 - i

    max = -1
    max_key = ' '
    for candidate, score in scores.items():
        if score > max:
            max = score
            max_key = candidate
    return max_key

def method_condorse(chains):
    scores = {c: 0 for c in chains[0]}

    for chain in chains:
        scores[chain[0]] += 1

    max_votes = -1
    winner = ' '
    for candidate, votes in scores.items():
        if votes > max_votes:
            max_votes = votes
            winner = candidate
    return winner

n, k = map(int, input().split())
chains = []
for i in range(k):
    print(f"Цепь для {i+1} избирателя:")
    chains.append(input().split())

print(f"По методу Борда выигрывает кандидат {method_bord(chains)}")
print(f"По методу Кондорсе выигрывает кандидат {method_condorse(chains)}")